File: Read-Write-Display-Image-OpenCV/image_io.py
#!/usr/bin/env python3
"""Read, optionally display, and write images safely with OpenCV."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2
import numpy as np


PROJECT_DIR = Path(__file__).resolve().parent
DEFAULT_INPUT = PROJECT_DIR / "assets" / "sample-scene.png"
READ_MODES = {
    "color": cv2.IMREAD_COLOR,
    "grayscale": cv2.IMREAD_GRAYSCALE,
    "unchanged": cv2.IMREAD_UNCHANGED,
}


def read_image(path: Path, mode: str = "color") -> np.ndarray:
    if mode not in READ_MODES:
        raise ValueError(f"mode must be one of {', '.join(READ_MODES)}")
    image = cv2.imread(str(path), READ_MODES[mode])
    if image is None:
        raise FileNotFoundError(f"Could not decode image: {path}")
    return image


def write_image(path: Path, image: np.ndarray, parameters: list[int] | None = None) -> None:
    if image.size == 0:
        raise ValueError("refusing to write an empty image")
    path.parent.mkdir(parents=True, exist_ok=True)
    if not cv2.imwrite(str(path), image, parameters or []):
        raise RuntimeError(f"Could not encode image: {path}")


def make_comparison(color: np.ndarray, gray: np.ndarray) -> np.ndarray:
    gray_bgr = cv2.cvtColor(gray, cv2.COLOR_GRAY2BGR)
    panels = [color.copy(), gray_bgr]
    for panel, label in zip(panels, ["IMREAD_COLOR", "IMREAD_GRAYSCALE"]):
        cv2.rectangle(panel, (0, 0), (panel.shape[1], 42), (0, 0, 0), -1)
        cv2.putText(
            panel,
            label,
            (14, 29),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.72,
            (255, 255, 255),
            2,
            cv2.LINE_AA,
        )
    return cv2.hconcat(panels)


def run_image_io(input_path: Path, output_dir: Path) -> dict[str, float | int]:
    color = read_image(input_path, "color")
    gray = read_image(input_path, "grayscale")
    unchanged = read_image(input_path, "unchanged")

    png_path = output_dir / "lossless-copy.png"
    gray_path = output_dir / "grayscale.png"
    jpeg_path = output_dir / "quality-90.jpg"
    comparison_path = output_dir / "image-io-comparison.png"
    write_image(png_path, color, [cv2.IMWRITE_PNG_COMPRESSION, 6])
    write_image(gray_path, gray)
    write_image(jpeg_path, color, [cv2.IMWRITE_JPEG_QUALITY, 90])
    write_image(comparison_path, make_comparison(color, gray))

    png_roundtrip = read_image(png_path, "color")
    jpeg_roundtrip = read_image(jpeg_path, "color")
    if not np.array_equal(color, png_roundtrip):
        raise AssertionError("PNG round trip unexpectedly changed pixel values")
    jpeg_mae = float(
        np.mean(np.abs(color.astype(np.int16) - jpeg_roundtrip.astype(np.int16)))
    )
    return {
        "width": color.shape[1],
        "height": color.shape[0],
        "color_channels": color.shape[2],
        "unchanged_channels": 1 if unchanged.ndim == 2 else unchanged.shape[2],
        "jpeg_mae": jpeg_mae,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=PROJECT_DIR / "outputs")
    parser.add_argument("--display", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    metrics = run_image_io(args.input, args.output_dir)
    print(
        f"size={metrics['width']}x{metrics['height']} "
        f"channels={metrics['color_channels']} "
        f"jpeg_mae={metrics['jpeg_mae']:.3f}"
    )
    if args.display:
        color = read_image(args.input, "color")
        gray = read_image(args.input, "grayscale")
        cv2.imshow("Read and display images", make_comparison(color, gray))
        cv2.waitKey(0)
        cv2.destroyAllWindows()


if __name__ == "__main__":
    main()

File: Read-Write-Display-Image-OpenCV/image_io.cpp
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace fs = std::filesystem;

cv::Mat readImage(const fs::path& path, int flags) {
    const cv::Mat image = cv::imread(path.string(), flags);
    if (image.empty()) throw std::runtime_error("could not decode " + path.string());
    return image;
}

void writeImage(const fs::path& path, const cv::Mat& image,
                const std::vector<int>& parameters = {}) {
    if (image.empty()) throw std::invalid_argument("refusing to write an empty image");
    fs::create_directories(path.parent_path());
    if (!cv::imwrite(path.string(), image, parameters)) {
        throw std::runtime_error("could not encode " + path.string());
    }
}

cv::Mat makeComparison(const cv::Mat& color, const cv::Mat& gray) {
    cv::Mat grayBgr;
    cv::cvtColor(gray, grayBgr, cv::COLOR_GRAY2BGR);
    std::vector<cv::Mat> panels{color.clone(), grayBgr};
    const std::vector<std::string> labels{"IMREAD_COLOR", "IMREAD_GRAYSCALE"};
    for (std::size_t i = 0; i < panels.size(); ++i) {
        cv::rectangle(panels[i], cv::Rect(0, 0, panels[i].cols, 42), cv::Scalar::all(0), -1);
        cv::putText(panels[i], labels[i], cv::Point(14, 29), cv::FONT_HERSHEY_SIMPLEX,
                    0.72, cv::Scalar::all(255), 2, cv::LINE_AA);
    }
    cv::Mat comparison;
    cv::hconcat(panels, comparison);
    return comparison;
}

double runImageIo(const fs::path& input, const fs::path& outputDir) {
    const cv::Mat color = readImage(input, cv::IMREAD_COLOR);
    const cv::Mat gray = readImage(input, cv::IMREAD_GRAYSCALE);
    const cv::Mat unchanged = readImage(input, cv::IMREAD_UNCHANGED);

    const fs::path pngPath = outputDir / "lossless-copy.png";
    const fs::path jpegPath = outputDir / "quality-90.jpg";
    writeImage(pngPath, color, {cv::IMWRITE_PNG_COMPRESSION, 6});
    writeImage(outputDir / "grayscale.png", gray);
    writeImage(jpegPath, color, {cv::IMWRITE_JPEG_QUALITY, 90});
    writeImage(outputDir / "image-io-comparison.png", makeComparison(color, gray));

    const cv::Mat pngRoundtrip = readImage(pngPath, cv::IMREAD_COLOR);
    cv::Mat pngDifference;
    cv::absdiff(color, pngRoundtrip, pngDifference);
    if (cv::countNonZero(pngDifference.reshape(1)) != 0) {
        throw std::runtime_error("PNG round trip unexpectedly changed pixel values");
    }

    const cv::Mat jpegRoundtrip = readImage(jpegPath, cv::IMREAD_COLOR);
    cv::Mat jpegDifference;
    cv::absdiff(color, jpegRoundtrip, jpegDifference);
    const double jpegMae = cv::mean(jpegDifference)[0] / 3.0 +
                           cv::mean(jpegDifference)[1] / 3.0 +
                           cv::mean(jpegDifference)[2] / 3.0;
    std::cout << "size=" << color.cols << "x" << color.rows
              << " channels=" << color.channels()
              << " unchanged_channels=" << unchanged.channels()
              << " jpeg_mae=" << jpegMae << '\n';
    return jpegMae;
}

int selfTest() {
    const fs::path directory = fs::current_path() / "image-io-self-test";
    const fs::path input = directory / "fixture.png";
    cv::Mat fixture(24, 32, CV_8UC3);
    for (int row = 0; row < fixture.rows; ++row) {
        for (int col = 0; col < fixture.cols; ++col) {
            fixture.at<cv::Vec3b>(row, col) =
                cv::Vec3b(static_cast<unsigned char>(col * 7),
                          static_cast<unsigned char>(row * 9),
                          static_cast<unsigned char>((row + col) * 4));
        }
    }
    writeImage(input, fixture);
    const double jpegMae = runImageIo(input, directory / "outputs");
    const bool passed = jpegMae >= 0.0 && jpegMae < 15.0;
    fs::remove_all(directory);
    if (!passed) {
        std::cerr << "self-test failed: jpeg_mae=" << jpegMae << '\n';
        return 1;
    }
    std::cout << "self-test passed\n";
    return 0;
}

int main(int argc, char** argv) {
    try {
        fs::path input = fs::path(__FILE__).parent_path() / "assets" / "sample-scene.png";
        fs::path outputDir = fs::path(__FILE__).parent_path() / "outputs";
        bool display = false;
        for (int i = 1; i < argc; ++i) {
            const std::string arg = argv[i];
            if (arg == "--self-test") return selfTest();
            if (arg == "--display") display = true;
            else if (arg == "--input" && i + 1 < argc) input = argv[++i];
            else if (arg == "--output-dir" && i + 1 < argc) outputDir = argv[++i];
            else if (arg != "--display") throw std::invalid_argument("unknown or incomplete option: " + arg);
        }
        runImageIo(input, outputDir);
        if (display) {
            const cv::Mat color = readImage(input, cv::IMREAD_COLOR);
            const cv::Mat gray = readImage(input, cv::IMREAD_GRAYSCALE);
            cv::imshow("Read and display images", makeComparison(color, gray));
            cv::waitKey(0);
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 1;
    }
}
